File: ruby/matrix/matrix.rb
class Matrix
  attr_reader :rows, :columns

  def initialize(string)
    @rows = generate_rows(string).freeze
    @columns = rows.transpose.freeze
  end

  private

  def generate_rows(string)
    string.each_line.map { |row| row.split.map(&:to_i) }
  end
end

File: javascript/matrix/matrix.js
export class Matrix {
  constructor(matrixString) {
    this._rows = this.generateRows(matrixString)
    this._columns = this.transpose(this.rows)
  }

  get rows() {
    return this._rows
  }

  get columns() {
    return this._columns
  }

  generateRows(matrixString) {
    return matrixString.split("\n").map(this.parseRowColumns)
  }

  parseRowColumns(rowString) {
    return rowString.split(" ").map(Number)
  }

  transpose(rows) {
    return (
      rows[0].map((_column, index) =>
        rows.map(row => row[index])
      )
    )
  }
}
